crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -330,34 +333,46 @@ impl<'db> NarrowingConstraintsBuilder<'db> {
         let scope = self.scope();
         let inference = infer_expression_types(self.db, expression);
 
-        if let Some(func_type) = inference
+        // TODO: add support for PEP 604 union types on the right hand side of `isinstance`
+        // and `issubclass`, for example `isinstance(x, str | (int | float))`.
+        match inference
             .expression_ty(expr_call.func.scoped_ast_id(self.db, scope))
             .into_function_literal()
+            .and_then(|f| f.known(self.db))
         {
-            if func_type.is_known(self.db, KnownFunction::IsInstance)
-                && expr_call.arguments.keywords.is_empty()
+            Some(function @ (KnownFunction::IsInstance | KnownFunction::IsSubclass))
+                if expr_call.arguments.keywords.is_empty() =>
             {
-                if let [ast::Expr::Name(ast::ExprName { id, .. }), rhs] = &*expr_call.arguments.args
+                if let [ast::Expr::Name(ast::ExprName { id, .. }), class_info] =
+                    &*expr_call.arguments.args
                 {
                     let symbol = self.symbols().symbol_id_by_name(id).unwrap();
 
-                    let rhs_type = inference.expression_ty(rhs.scoped_ast_id(self.db, scope));
+                    let class_info_ty =
+                        inference.expression_ty(class_info.scoped_ast_id(self.db, scope));
 
-                    // TODO: add support for PEP 604 union types on the right hand side:
-                    // isinstance(x, str | (int | float))
-                    if let Some(mut constraint) = generate_isinstance_constraint(self.db, &rhs_type)
-                    {
-                        if !is_positive {
-                            constraint = constraint.negate(self.db);
-                        }
-                        let mut constraints = NarrowingConstraints::default();
-                        constraints.insert(symbol, constraint);
-                        return Some(constraints);
-                    }
+                    #[allow(clippy::match_wildcard_for_single_variants)]
+                    let to_constraint = match function {
+                        KnownFunction::IsInstance => |class_literal: ClassLiteralType<'db>| {
+                            Type::anonymous_instance(class_literal.class)
+                        },
+                        KnownFunction::IsSubclass => Type::Type,
+                        _ => unreachable!(),
+                    };
```

Comment:
I think you could avoid the Clippy supppression and the `unreachable!()` here if you did something like this:

<details>
<summary>Suggested revision</summary>

```rs
        // TODO: add support for PEP 604 union types on the right hand side of `isinstance`
        // and `issubclass`, for example `isinstance(x, str | (int | float))`.
        let constraint_function = inference
            .expression_ty(expr_call.func.scoped_ast_id(self.db, scope))
            .into_function_literal()
            .and_then(|f| f.known(self.db))
            .and_then(KnownConstraintFunction::try_from_known_function)?;

        if expr_call.arguments.keywords.is_empty() {
            if let [ast::Expr::Name(ast::ExprName { id, .. }), class_info] =
                &*expr_call.arguments.args
            {
                let symbol = self.symbols().symbol_id_by_name(id).unwrap();

                let class_info_ty =
                    inference.expression_ty(class_info.scoped_ast_id(self.db, scope));

                let to_constraint = match constraint_function {
                    KnownConstraintFunction::IsInstance => {
                        |class_literal: ClassLiteralType<'db>| {
                            Type::anonymous_instance(class_literal.class)
                        }
                    }
                    KnownConstraintFunction::IsSubclass => Type::Type,
                };

                generate_classinfo_constraint(self.db, &class_info_ty, to_constraint).map(
                    |constraint| {
                        let mut constraints = NarrowingConstraints::default();
                        constraints.insert(symbol, constraint.negate_if(self.db, !is_positive));
                        constraints
                    },
                )
            } else {
                None
            }
        } else {
            None
        }
```

</details>

Where `KnownConstraintFunction` is an enum defined like this:

<details>

```rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum KnownConstraintFunction {
    IsInstance,
    IsSubclass,
}

impl KnownConstraintFunction {
    const fn try_from_known_function(function: KnownFunction) -> Option<Self> {
        match function {
            KnownFunction::IsSubclass => Some(Self::IsSubclass),
            KnownFunction::IsInstance => Some(Self::IsInstance),
            KnownFunction::RevealType => None,
        }
    }
}
```

</details>

Reply:
I modelled it slightly differently using:
```rs
pub enum KnownFunction {
    ConstraintFunction(KnownConstraintFunction),
    RevealType,
}
```